test/unit/transform/classes_test.py

```diff
@@ -0,0 +1,791 @@
+            check_names=False,
+        )
+
+
+@pytest.mark.parametrize(
+    "df,expected,params",
+    [
+        pytest.param(
+            NUMERIC_DATA,
+            NUMERIC_DATA.loc[NUMERIC_DATA.id.isin([1, 2, 3, 4, 5, 6, 8, 9, 10])],
+            dict(
+                invalid_values=[0, pd.NA, np.nan],
+                required_valid_cols=[
+                    "valid_year",
+                    "valid_capacity_mw",
+                    "net_generation_mwh",
+                ],
+            ),
+            id="required_valid_cols",
+        ),
+        pytest.param(
+            NUMERIC_DATA,
+            NUMERIC_DATA.loc[NUMERIC_DATA.id.isin([1, 2, 3, 4, 5, 6, 8, 9, 10])],
+            dict(
+                invalid_values=[0, pd.NA, np.nan],
+                allowed_invalid_cols=[
+                    "id",
+                    "year",
+                    "capacity_kw",
+                    "capacity_mw",
+                    "net_generation_kwh",
+                ],
+            ),
+            id="allowed_invalid_cols",
+        ),
+    ],
+)
+def test_drop_invalid_rows(df, expected, params):
+    """Test our ability to select and drop invalid rows."""
+    invalid_row = InvalidRows(**params)
+    actual = drop_invalid_rows(df, params=invalid_row)
+    assert_frame_equal(actual, expected)
+
+
+#####################################################################################
+# TableTransformer class unit tests.
+#####################################################################################
+@enum.unique
+class TableId(enum.Enum):
+    """Dummy test table IDs."""
+
+    TEST_TABLE = "test_table"
+
+
+class TableTransformer(AbstractTableTransformer):
+    """A concrete TableTransformer for testing purposes."""
+
+    table_id: enum.Enum = TableId.TEST_TABLE
+
+    @cache_df(key="start")
+    def transform_start(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Start the transform."""
+        df = self.rename_columns(df).pipe(self.convert_units)
+        return df
+
+    @cache_df(key="main")
+    def transform_main(self, df: pd.DataFrame) -> pd.DataFrame:
+        """The main body of the transform."""
+        df = (
+            self.normalize_strings(df)
+            .pipe(self.categorize_strings)
+            .pipe(self.nullify_outliers)
+        )
+        return df
+
+    @cache_df(key="end")
+    def transform_end(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Finish up the transform."""
+        df = self.drop_invalid_rows(df).convert_dtypes(convert_floating=False)
+        return df
+
+
+@pytest.mark.parametrize(
+    "df,params",
+    [
+        pytest.param(STRING_DATA, STRING_PARAMS, id="string_transform"),
+        pytest.param(NUMERIC_DATA, NUMERIC_PARAMS, id="numeric_transform"),
+    ],
+)
+def test_transform(df, params):
+    """Test the use of general transforms as part of a TableTransfomer class.
+
+    Should really define "expected" output dataframes here and check that they match.
```

Comment:
I hesitated to do this because the various unit tests were starting to feel too coupled to each other -- such that adding a new test case to one would end up changing expectations elsewhere, since they're sharing input / output data. Do you have any thoughts on how to keep them better comparmentalized?

Reply:
What I might do for something like this is just test that the functions like `normalize_strings_mutlicol` is called with the proper inputs, but not actually test the underlying function since that is being done in other tests. You could do this with [pytest-mock](https://pytest-mock.readthedocs.io/en/latest/usage.html), which I used for tests in the XBRL scraper. This might look something like:
```py
def test_transform(df, params, mocker):
    normalize_strings_mock = mocker.MagickMock(name="normalize_strings", return_value=df)
    mocker.patch("pudl.transform.classes.normalize_strings_multicol", new=nromalize_strings_mock)

    transformer = TableTransformer(
        params=TableTransformParams.from_dict(params["test_table"]),
        cache_dfs=True,
        clear_cached_dfs=False,
    )

    normalize_strings_mock.assert_called_once_with(df, params)
```
This would replace the actual function with the mock that just records what it was called with and returns the original dataframe. This way you're only testing that the proper functions are called and not what they do since that's not really the `AbstractTableTransformer`'s responsibility.

You could also create another test that tests the caching by doing some pointless operation in one of the phases of the transform, like set all values in the dataframe to 15, then check that the appropriate dataframe is in the cache.

Reply:
Okay @zschira  I've updated the `TableTransformer` test to mock the internal functions but something weird is going on, or I don't understand how to do the mocking -- even though I've got the mock functions set up and they seem to be working (I can access their attributes after they've run) the mock functions don't appear to be the ones getting run -- instead the original functions are getting called and the dataframe is being mutated, so the comparison of the input values isn't working.

I could just check whether they were getting called with any dataframe at all and not worry about why it's not the dataframe I expected, but I feel like there must be something I'm not understanding.

Reply:
Is that because they're getting called from within class methods? Or because they aren't normal "functions" and rather are `Callable` objects that were constructed by the `multicol_transform_factory()` function??

Reply:
Mocking `drop_invalid_rows()` which isn't a constructed `Callable` behaves as expected, so I do think that's the root of the issue here.

If I try to mock the `__call__()` method specifically, then it doesn't show up as having been called at all (it's called zero times, and has no `args` to compare).

```py
    mocker.patch(
        "pudl.transform.classes.nullify_outliers_multicol.__call__",
        new=nullify_outliers_mock,
    )
```

Reply:
I found a [stackoverflow post](https://stackoverflow.com/questions/61699871/mock-function-that-had-decorator-use-same-decorator-again-to-decorate-mock-obje) that seemed to address a similar problem, and I played around with it for a bit, but I can't get it working. I wonder if this is just too complicated with wrapped functions like this, and we should try a different path?

Reply:
I simplified the test to just check that the argument was a dataframe, and not the contents of the dataframe so that @cmgosnell could pull the changes over into her utility ID branch. I kinda feel like this is good enough and we should maybe move on. Take a look and let me know what you think.

Reply:
This seems good enough to me. I don't think this test is really valuable enough to spend more time on honestly